sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala

```diff
@@ -485,6 +485,28 @@ case class Literal (value: Any, dataType: DataType) extends LeafExpression {
       toDayTimeIntervalString(i, ANSI_STYLE, startField, endField)
     case (i: Int, YearMonthIntervalType(startField, endField)) =>
       toYearMonthIntervalString(i, ANSI_STYLE, startField, endField)
+    case (data: GenericArrayData, arrayType: ArrayType) =>
+      val arrayValues: Array[String] =
+        data.array.map {
+          Literal(_, arrayType.elementType).sql
+        }
+      s"ARRAY(${arrayValues.mkString(", ")})"
+    case (row: GenericInternalRow, structType: StructType) =>
+      val structValues: Array[String] =
+        row.values.zip(structType.fields.map(_.dataType)).map {
+          case (value: Any, fieldType: DataType) =>
+            Literal(value, fieldType).sql
+        }
+      s"STRUCT(${structValues.mkString(", ")})"
```

Comment:
probably using a named structure is better, otherwise the column name of the literal is always "col1","col2"..